### DISCUSSION TITLE
igl_copy_dll does not copy the dll files I need

### DISCUSSION THREAD

**Discussion by Jarelk** [libigl/libigl#2091] [Q&A]
Hello, I am trying to get the example project working with CGAL, Embree, and Tetgen. Everything is going okay, except that I need to manually copy the GMP/MPFR dll files to the executable folder. The `igl_copy_dll` cmake function does not seem to work.

Below is my CMakeLists.txt.

The `cmake` folder in my project directory contains `libigl.cmake` and `options.cmake`, which don't do much except fetch Libigl and set the cmake options for Libigl.
```cmake
cmake_minimum_required(VERSION 3.16)
project(INFOMR)

set (CMAKE_CXX_STANDARD 17)
set_property(GLOBAL PROPERTY USE_FOLDERS ON)
list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)

# Options
include(options)

# Libigl
include(libigl)

# Enable the target igl::glfw
igl_include(glfw)

# Add your project files
file(GLOB_RECURSE SOURCE CONFIGURE_DEPENDS "src/*.cpp" "src/*.c")
file(GLOB_RECURSE HEADERS CONFIGURE_DEPENDS "src/*.hpp" "src/*.h")

add_executable(${PROJECT_NAME} ${SOURCE} ${HEADERS})
target_link_libraries(${PROJECT_NAME} PUBLIC igl::glfw igl_copyleft::cgal igl_copyleft::tetgen igl::embree)
igl_copy_dll(${PROJECT_NAME})

set_target_properties(INFOMR PROPERTIES COMPILE_FLAGS "/GL")
set_target_properties(INFOMR PROPERTIES LINK_FLAGS "/LTCG")
target_precompile_headers(INFOMR PRIVATE src/precomp.hpp)```

**Comment by jdumas** [comment]
Can you share the content of your `<build>/runtime_deps` folder if you have anything in there?

**Reply by Jarelk** [reply]
I do. Just three 0 kB cmake files. They are named by the function with `"${CMAKE_BINARY_DIR}/runtime_deps/copy_dll_${target}_$<CONFIG>.cmake"`. One for Debug, one for Release, one for "RelWithDebInfo" whatever that is.

**Reply by jdumas** [reply]
Interesting. Can you confirm whether you get the same behavior if you try to build the libigl repo by itself (i.e. not just using it in another project)?

**Reply by jdumas** [reply]
Alternatively, it'd be great if you could help diagnose what's wrong with the current function, e.g. by inserting a bunch of `message()` . For example, what is the value of `TARGET_DEPENDENCIES` after this line:
https://github.com/libigl/libigl/blob/574ab1a3a819b3798cf452fa44bcb47704a529c6/cmake/igl/igl_copy_dll.cmake#L71
Or what are the dependencies being iterated on in this line:
https://github.com/libigl/libigl/blob/574ab1a3a819b3798cf452fa44bcb47704a529c6/cmake/igl/igl_copy_dll.cmake#L23

My Windows machine is undergoing repairs theses days so I can't really test this myself.

**Reply by Jarelk** [reply]
Here are the dependencies being iterated. As for building the libigl repo, do you mean the example project, or the library on its own?

```Creating target: igl::core (igl_core)
Third-party: creating target 'Eigen3::Eigen'
Creating target: igl::embree (igl_embree)
Third-party: creating target 'embree::embree'
Detecting default ISA...
Detected default ISA: SSE2
Creating target: igl::opengl (igl_opengl)
Third-party: creating target 'glad::glad'
Creating target: igl::glfw (igl_glfw)
Third-party: creating target 'glfw::glfw'
Using Win32 for window creation
Creating target: igl::imgui (igl_imgui)
Third-party: creating target 'imgui::imgui'
Third-party: creating target 'imguizmo::imguizmo'
Third-party: creating target 'igl::imgui_fonts'
Creating target: igl_copyleft::core (igl_copyleft_core)
Creating target: igl_copyleft::cgal (igl_copyleft_cgal)
Third-party: creating target 'CGAL::CGAL'
Third-party: downloading gmp + mpfr
Third-party: creating target 'gmp::gmp'
Third-party: downloading gmp + mpfr
Third-party: creating target 'mpfr::mpfr'
Third-party: creating targets 'Boost::boost'...
Fetching Boost
Fetching Boost - done
Boost found: 1.71.0 K:/Informatica/Multimedia_Retrieval/INFOMR/build/_deps/boost-src
Could NOT find BZip2 (missing: BZIP2_LIBRARIES BZIP2_INCLUDE_DIR) 
Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR) 
Using header-only CGAL
Boost include dirs: 
Boost libraries:    Boost::thread;Boost::system
Creating target: igl_copyleft::tetgen (igl_copyleft_tetgen)
Third-party: creating target 'tetgen::tetgen'
Current dependency: igl::glfw
Current dependency: igl::core
Current dependency: Eigen3::Eigen
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: Threads::Threads
Current dependency: igl::opengl
Current dependency: igl::core
Current dependency: glad::glad
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: $<IF:$<TARGET_EXISTS:OpenGL::OpenGL>,OpenGL::OpenGL,OpenGL::GL>
Current dependency: glfw::glfw
Current dependency: Threads::Threads
Current dependency: igl_copyleft::cgal
Current dependency: igl::core
Current dependency: igl_copyleft::core
Current dependency: igl::core
Current dependency: CGAL::CGAL
Current dependency: gmp::gmp
Current dependency: mpfr::mpfr
Current dependency: Boost::thread
Current dependency: Boost::boost
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: Boost::chrono
Current dependency: Boost::boost
Current dependency: Boost::system
Current dependency: Boost::boost
Current dependency: CGAL::CGAL_Core
Current dependency: gmp::gmp
Current dependency: mpfr::mpfr
Current dependency: CGAL::CGAL
Current dependency: igl_copyleft::tetgen
Current dependency: igl::core
Current dependency: igl_copyleft::core
Current dependency: tetgen::tetgen
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: igl::embree
Current dependency: igl::core
Current dependency: embree::embree
Current dependency: sys
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: math
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: simd
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
Current dependency: lexers
Current dependency: sys
Current dependency: math
Current dependency: tasking
Current dependency: TARGET_DEPENDENCIES-NOTFOUND
All dependencies: igl_glfw;igl_core;Eigen3_Eigen;igl_opengl;glad;glfw;igl_copyleft_cgal;igl_copyleft_core;CGAL;Boost_thread;Boost::boost;Boost_chrono;Boost_system;CGAL_Core;igl_copyleft_tetgen;tetgen;igl_embree;embree;sys;math;simd;lexers;tasking
Populating copy rules for target: INFOMR
Configuring done
```

**Reply by jdumas** [reply]
> As for building the libigl repo, do you mean the example project, or the library on its own?

The library on its own, but if you can try both it'd be great.

**Reply by Jarelk** [reply]
What am I supposed to be looking for in building the library itself? I am not sure which executable, if any, I should pass as target to the function.

Edit: I should clarify, CMake is black magic to me.

**Reply by Jarelk** [reply]
Can confirm the example project also just makes a bunch of 0 kB cmake files in the `runtime_deps` folder.

**Reply by jdumas** [reply]
When you build the library you should have a number of executable for each tutorial in the `<build>/bin` folder. It sounds like you won't be able to run them due to missing dll either. Can you tell me more about how you're trying to build with CMake? Which version of Visual Studio are you using? Are you generating the VS project via CMake-GUI, or are you trying to open the CMake file directly from Visual Studio?

**Reply by Jarelk** [reply]
Hello, sorry for the late reply. I have been busy doing uni stuff (with Libigl)! I haven't had time to build the libigl tutorials yet. I can tell you about my Cmake however: I'm using the CMake GUI latest version, and Visual Studio 2022. I'm generating the VS Project via CMake.
